runtime/vm/vmbootlib.c

```diff
@@ -582,10 +584,10 @@ classLoaderRegisterLibrary(void *voidVMThread, J9ClassLoader *classLoader, const
 		if (NULL == newNativeLibrary->logicalName) {
 			reportError(errBuf, "Failed allocating memory for native library logical name", bufLen);
 			rc = J9NATIVELIB_LOAD_ERR_OUT_OF_MEMORY;
+			goto leave_routine;
```

Comment:
Seems this should free the newNativeLibrary->name first.

Reply:
Also also free the newNativeLibrary pool element, here an the previous goto.

Reply:
There is a freeSharedLibrary() helper which is used below, but not after leave_routine

Reply:
Moved `freeSharedLibrary()` below `leave_routine` for cleanup.

Reply:
I don't like that because it changes the existing behavior. There are other places which `goto leave_routine` that may no longer be correct.

Reply:
Also freeSharedLibrary() calls `(*library->close)(vmThread, library);` without checking things are setup properly to do that, which may cause a crash.

Reply:
I think existing two `goto leave_routine` missed the cleanup performed by `freeSharedLibrary()` which will be fixed by this PR.

Reply:
I'm not sure, it's possible the entries for the library which couldn't be opened are supposed to remain in the table, in particular for the jniVersion mismatch case. We can work on figuring it out before merging this, or leave it for a future PR. What you have now has the `(*library->close)(vmThread, library);` problem, so at least that needs to be fixed. I expect `*library->close` will be NULL before initializeNativeLibrary() is called.

Reply:
The `newNativeLibrary` is put into `classLoader->librariesHead` & `librariesTail` only when `J9NATIVELIB_LOAD_OK == rc` before returning to the caller. It might be set via `*libraryPtr = newNativeLibrary` however `libraryPtr` shouldn't be used if `J9NATIVELIB_LOAD_OK != rc` which is the case when `goto leave_routine`.

It is correct that `*library->close` will be `NULL` before `initializeNativeLibrary()` is called. Note that `initializeNativeLibrary()` is invoked right after `newNativeLibrary` is allocated successfully, first `goto leave_routine` occurs when `NULL == newNativeLibrary` which won't pass the condition to call `freeSharedLibrary()`, other `goto leave_routine` only occur after  `initializeNativeLibrary()`.
I don't see a problem.

Reply:
ok